module/zfs/zfs_ctldir.c

```diff
@@ -809,32 +947,46 @@ zfsctl_unmount_snapshots(zfs_sb_t *zsb, int flags, int *count)
 	"mount -t zfs -n '%s' '%s'"
 
 int
-zfsctl_mount_snapshot(struct path *path, int flags)
+zfsctl_snapshot_mount(struct path *path, int flags)
 {
 	struct dentry *dentry = path->dentry;
 	struct inode *ip = dentry->d_inode;
-	zfs_sb_t *zsb = ITOZSB(ip);
+	zfs_sb_t *zsb;
+	zfs_sb_t *snap_zsb;
+	zfs_snapentry_t *se;
 	char *full_name, *full_path;
-	zfs_snapentry_t *sep;
-	zfs_snapentry_t search;
 	char *argv[] = { "/bin/sh", "-c", NULL, NULL };
 	char *envp[] = { NULL };
 	int error;
 
+	if (ip == NULL)
+		return (EISDIR);
+
+	zsb = ITOZSB(ip);
 	ZFS_ENTER(zsb);
 
 	full_name = kmem_zalloc(MAXNAMELEN, KM_SLEEP);
-	full_path = kmem_zalloc(PATH_MAX, KM_SLEEP);
+	full_path = kmem_zalloc(MAXPATHLEN, KM_SLEEP);
 
-	error = zfsctl_snapshot_zname(ip, dname(dentry), MAXNAMELEN, full_name);
+	error = zfsctl_snapshot_name(zsb, dname(dentry),
+	    MAXNAMELEN, full_name);
 	if (error)
 		goto error;
 
-	error = zfsctl_snapshot_zpath(path, PATH_MAX, full_path);
+	error = zfsctl_snapshot_path(path, MAXPATHLEN, full_path);
 	if (error)
 		goto error;
 
 	/*
+	 * Multiple concurrent automounts of a snapshot are never allowed.
+	 * The snapshot may be manually mounted as many times as desired.
+	 */
+	if (zfsctl_snapshot_ismounted(full_name)) {
```

Comment:
Right, a race is still possible here since a lock isn't held over the ismount check and subsequent unmount.  However, if this happens there are no negative consequences.  The check is there mainly to just keep this window as small as possible.